### PR TITLE
Ensure that LoadSequentialFile() actually read the whole file

### LINKED ISSUE #5826: Error while loading model
After train a model for a long time I saved it using

`bst.save_model("xgboostModel.trail04.json")`

I latter tried to load it with the same params to continue training 
```
filemodel = 'xgboostModel.trail04.json'
bst = xgb.train(param, xgTrain, num_boost_round=numOfRounds, evals=watchList, early_stopping_rounds=2000, xgb_model=filemodel)
```
but I get this error:

```
  File "/home/user/programs/pycharm/plugins/python-ce/helpers/pydev/pydevd.py", line 1438, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/home/user/programs/pycharm/plugins/python-ce/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/user/projects/temp/pythontmp/lastPointClassifiers.py", line 270, in <module>
    trial04()
  File "/home/user/projects/temp/pythontmp/lastPointClassifiers.py", line 256, in trial04
    xgb_model=filemodel)
  File "/home/user/.local/lib/python3.7/site-packages/xgboost/training.py", line 212, in train
    xgb_model=xgb_model, callbacks=callbacks)
  File "/home/user/.local/lib/python3.7/site-packages/xgboost/training.py", line 37, in _train_internal
    model_file=xgb_model)
  File "/home/user/.local/lib/python3.7/site-packages/xgboost/core.py", line 1175, in __init__
    self.load_model(model_file)
  File "/home/user/.local/lib/python3.7/site-packages/xgboost/core.py", line 1804, in load_model
    self.handle, c_str(os_fspath(fname))))
  File "/home/user/.local/lib/python3.7/site-packages/xgboost/core.py", line 190, in _check_call
    raise XGBoostError(py_str(_LIB.XGBGetLastError()))
xgboost.core.XGBoostError: [19:57:42] /workspace/src/common/json.cc:413: Expecting: ",", got: "

```

When I try the same exact thing but run it over a smaller number of rounds it loads fine and trains like a champ.

```
def Example():
    X, Y = getData(randomize=True)
    X = np.array(X)
    Y = np.array(Y)
    l = int(len(X) * .5)
    xgTrain = xgb.DMatrix(X[:l], label=Y[:l])
    xgTest = xgb.DMatrix(X[l:], label=Y[l:])
    param = {
        'objective': 'multi:softmax',
        'eta': 0.0001,
        'max_depth': 6,
        'nthread': 16,
        'num_class': 2,
        'tree_method': 'gpu_hist',
    }
    watchList = [(xgTrain, 'train'), (xgTest, 'test')]
    numOfRounds = 1000000
    filemodel = 'xgboostModel.trail04.json'

    print("fit()")
    if path.exists(filemodel):
        bst = xgb.train(param, xgTrain, num_boost_round=numOfRounds, evals=watchList, early_stopping_rounds=2000,
                        xgb_model=filemodel)
    else:
        bst = xgb.train(param, xgTrain, num_boost_round=numOfRounds, evals=watchList, early_stopping_rounds=2000)

    bst.save_model(filemodel)
```

### PR DESCRIPTION
**Pull request by hcho3**
Fixes #5826.

On Linux, XGBoost uses the POSIX function `read()` to read the JSON model file. The `read()` function may return fewer bytes than the number of bytes requested, according to the [POSIX standard](https://pubs.opengroup.org/onlinepubs/009695399/functions/read.html):

> The value returned may be less than _nbyte_ if the number of bytes left in the file is less than _nbyte_

Fix: Call `read()` multiple times until we actually read the whole file.

### CODE REVIEW

File: src/common/io.cc [hunk 1]
```diff
@@ -120,27 +121,22 @@ std::string LoadSequentialFile(std::string fname) {
...
+  buffer.back() = '\0';
 #else  // defined(__unix__)
-  FILE *f = fopen(fname.c_str(), "r");
-  if (f == NULL) {
-    std::string msg;
-    OpenErr();
-  }
-  fseek(f, 0, SEEK_END);
-  auto fsize = ftell(f);
-  fseek(f, 0, SEEK_SET);
```
Comment (hcho3) [inline src/common/io.cc:137]: Note: According to the [MSDN documentation](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/ftell-ftelli64?view=vs-2019), `ftell()` may not yield the correct byte offset when the file was opened in the text mode.

Fix. We do away with `fseek()` and `ftell()` entirely and use `std::istreambuf_iterator` to read the whole file into `std::string`. The fix only applies on Windows. On Linux, we still use `read()`, along with `stat()` to measure the file size.

File: src/common/io.cc [hunk 2]
```diff
@@ -120,27 +121,30 @@ std::string LoadSequentialFile(std::string fname) {
...
-
-  buffer.resize(fsize + 1);
-  fread(&buffer[0], 1, fsize, f);
-  fclose(f);
+  buffer = buffer.substr(0, i);
+#else  // defined(__unix__)
+  std::ifstream ifs(fname);
```
Comment (trivialfis) [inline src/common/io.cc:144]: Please note that this is terribly slow.

Comment (hcho3) [reply src/common/io.cc:144]: Slower than `std::fread()`?

Comment (hcho3) [reply src/common/io.cc:144]: In https://social.msdn.microsoft.com/Forums/en-US/87e128e2-5320-4796-b4a2-36248f86a8ce/huge-performance-hit-when-using-ifstream-vs-fopen?forum=Vsexpressvc, `ifstream` shows competitive performance with respect to `std::fread()`.

File: include/xgboost/json_io.h [hunk 3]
```diff
@@ -84,8 +85,10 @@ class JsonReader {
     std::string msg = "Expecting: \"";
     msg += c;
     msg += "\", got: \"";
-    if (got == -1) {
+    if (got == EOF) {
```
Comment (trivialfis) [inline include/xgboost/json_io.h:88]: Please don't remove `-1` case.

File: src/common/io.cc [hunk 4]
```diff
@@ -120,27 +121,30 @@ std::string LoadSequentialFile(std::string fname) {
...
+    bytes_read += result;
   }
   close(fd);
-#else  // defined(__unix__)
-  FILE *f = fopen(fname.c_str(), "r");
-  if (f == NULL) {
-    std::string msg;
-    OpenErr();
+  buffer.back() = '\0';
+  // Trim NUL letters at the end
```
Comment (trivialfis) [inline src/common/io.cc:135]: Where is the NULL inserted?

File: src/common/io.cc [hunk 5]
```diff
@@ -120,27 +121,30 @@ std::string LoadSequentialFile(std::string fname) {
...
+    }
   }
-  fseek(f, 0, SEEK_END);
-  auto fsize = ftell(f);
-  fseek(f, 0, SEEK_SET);
-
-  buffer.resize(fsize + 1);
-  fread(&buffer[0], 1, fsize, f);
-  fclose(f);
+  buffer = buffer.substr(0, i);
```
Comment (trivialfis) [inline src/common/io.cc:142]: Why is it a substr?

File: tests/cpp/common/test_io.cc [hunk 6]
```diff
@@ -39,5 +44,50 @@ TEST(IO, FixedSizeStream) {
     ASSERT_EQ(huge_buffer, out_buffer);
   }
 }
+
+
+#if SIZE_MAX == 0xFFFFFFFFFFFFFFFF  // Only run this test on 64-bit system
+TEST(IO, LoadSequentialFile) {
```
Comment (trivialfis) [inline tests/cpp/common/test_io.cc:50]: Could you please clarify the bug a bit more?  I'm confused.

> The value returned may be less than nbyte if the number of bytes left in the file is less than nbyte

Can we rephrase it to: maximum returned value is actual bytes left in the file being read?  If yes, then what's the condition to trigger the bug that we are trying to fix?  Since as long as we don't try to read extra non-existing bytes, the returned value should be correct no matter how large is the file right?  And here we estimate the file size by:
```
  size_t f_size_bytes = fs.st_size;
  buffer.resize(f_size_bytes + 1);
```
So, if `fs.st_size` is a correct value, then we are reading exactly the number of bytes in file. Why would we end up with an half loaded buffer?


Looking at the test, it seems the real cause is (quoting from the link you provided):

> If the value of nbyte is greater than {SSIZE_MAX}, the result is implementation-defined.

Can we add an optional parameter to the load function specifying batch size for test, instead of creating a huge file?

Comment (trivialfis) [reply tests/cpp/common/test_io.cc:50]: Also, as now we have replaced the number printing algorithm, generated file should be much much smaller.

File: include/xgboost/json_io.h (hunk 3 above)
Comment (hcho3) [reply include/xgboost/json_io.h:88]: No, on some systems, EOF is not necessarily -1.

File: tests/cpp/common/test_io.cc (hunk 6 above)
Comment (hcho3) [reply tests/cpp/common/test_io.cc:50]: The bug has nothing to do with SSIZE_MAX. The bug has to do with the behavior of read(). When we call read() function with nbytes=2.3GB, it does not actually read 2.3GB. So we must call read() twice or more to get all the bytes into the buffer; otherwise we end up with half full buffer.

File: src/common/io.cc (hunk 5 above)
Comment (hcho3) [reply src/common/io.cc:142]: My bad. It should be resize() so that no copy is made.

File: tests/cpp/common/test_io.cc (hunk 6 above)
Comment (hcho3) [reply tests/cpp/common/test_io.cc:50]: > Can we add an optional parameter to the load function specifying batch size for test, instead of creating a huge file?

Not sure what you mean here. How can we test the behavior of read() with nbytes=2.3GB without creating a tempfile of size 2.3GB?

File: include/xgboost/json_io.h (hunk 3 above)
Comment (hcho3) [reply include/xgboost/json_io.h:88]: Sorry, you are right. We are reading from a char array, so any EOF that appears would necessarily be -1. On ARM chips, EOF is actually bigger than char, so this case won’t apply.

File: src/common/io.cc (hunk 4 above)
Comment (hcho3) [reply src/common/io.cc:135]: Removed.

File: src/common/io.cc (hunk 5 above)
Comment (hcho3) [reply src/common/io.cc:142]: Never mind. I’ll remove this change.

File: tests/cpp/common/test_io.cc (hunk 6 above)
Comment (hcho3) [reply tests/cpp/common/test_io.cc:50]: > Can we rephrase it to: maximum returned value is actual bytes left in the file being read?

No. The standard lists a few possible scenarios where read() returns fewer bytes than the number of bytes requested (nbytes parameter). One possible scenario is to have a short file. For this particular bug, we do have a complete file, yet read() returns fewer bytes. I honestly don’t understand why read() gives fewer than 2.3GB.

This behavior is quite well known. The prevailing wisdom in UNIX community is to wrap read() in a while loop.

**Comment by hcho3** [comment]
@trivialfis I removed the extra changes, as you requested. Can I get another round of review?

File: src/common/io.cc [hunk 7]
```diff
@@ -120,27 +121,22 @@ std::string LoadSequentialFile(std::string fname) {
 #if defined(__linux__)
   posix_fadvise(fd, 0, 0, POSIX_FADV_SEQUENTIAL);
 #endif  // defined(__linux__)
-  ssize_t bytes_read = read(fd, &buffer[0], f_size_bytes);
-  if (bytes_read < 0) {
-    close(fd);
-    ReadErr();
+  ssize_t bytes_read = 0;
+  while (bytes_read < f_size_bytes) {
```
Comment (hcho3) [inline src/common/io.cc:125]: @trivialfis Take a look at this while loop. This is a commonly used construct in UNIX programs.

File: src/common/io.cc [hunk 8]
```diff
@@ -120,27 +121,22 @@ std::string LoadSequentialFile(std::string fname) {
 #if defined(__linux__)
   posix_fadvise(fd, 0, 0, POSIX_FADV_SEQUENTIAL);
 #endif  // defined(__linux__)
-  ssize_t bytes_read = read(fd, &buffer[0], f_size_bytes);
```
Comment (hcho3) [inline src/common/io.cc:123]: @trivialfis Previously, we called `read()` only once. For a file containing 2896075944 bytes (2.69 GB), `read()` only reads 2147479552 bytes (1.99 GB), and the variables would be set as follows:
```
bytes_read = 2147479552, f_size_bytes = 2896075944
```

This PR ensures that `bytes_read` would be identical to `f_size_bytes`.

File: src/common/io.cc (hunk 2 above)
Comment (trivialfis) [reply src/common/io.cc:136]: > std::ifstream shows competitive performance with respect to std::fread().

Interesting as I just tried on Linux, it's about 10x slower.  But I guess when file is that huge JSON would be the bottleneck here.

File: tests/cpp/common/test_io.cc (hunk 6 above)
Comment (trivialfis) [reply tests/cpp/common/test_io.cc:50]: > Not sure what you mean here. How can we test the behavior of read() with nbytes=2.69 GB without creating a tempfile of size 2.69 GB?

What I'm worrying about is the behaviour is dependent on other factors like available memory.  So 2.69 GB might not be a correct test.
